source/characterProcessing.py

```diff
@@ -779,9 +756,156 @@ def clearSpeechSymbols():
+		"""Gets the symbols for a given locale.
+		:param locale: The locale to get symbols for.
+		:raises FileNotFoundError: When this is not a user dictionary and the locale wasn't found.
+		"""
+		return self.symbols.fetchLocaleData(locale, fallback=False)
+
+	def _getSymbols(self, locale: str) -> SpeechSymbols:
+		raiseOnError = not self.user
+		symbols = SpeechSymbols()
+		if locale not in self.availableLocales:
+			msg = f"No {self.name!r} data for locale {locale!r}"
+			if raiseOnError:
+				raise FileNotFoundError(msg)
+			log.debug(msg)
+		else:
+			try:
+				symbols.load(self.path.format(locale=locale), self.allowComplexSymbols)
+			except IOError:
+				if raiseOnError:
+					raise
+				log.error(f"Error loading {self.name!r} data for locale {locale!r}", exc_info=True)
+		return symbols
+
+	@cached_property
+	def availableLocales(self) -> dict[str, str]:
+		"""Gets dictionary paths for all available locales."""
+		prefix, suffix = self.path.split("{locale}", 1)
+		pattern = f"{prefix}*{suffix}"
+		paths = glob.glob(pattern)
+		dct = {}
+		for p in paths:
+			locale = p[len(prefix) : (-1 * len(suffix))]
+			dct[locale] = p
+		return dct
+
+
+_symbolDictionaryDefinitions: list[SymbolDictionaryDefinition] = []
+
+
+def getAvailableSymbolDictionaryDefinitions() -> Generator[SymbolDictionaryDefinition, None, None]:
+	"""Get available symbol dictionary definitions as initialized in core or in add-ons."""
+	yield from _symbolDictionaryDefinitions
+
+
+def _addSymbolDefinitions():
+	# Add symbol dictionary definitions
+	_symbolDictionaryDefinitions.append(
+		SymbolDictionaryDefinition(
+			name="cldr",
+			path=os.path.join(globalVars.appDir, "locale", "{locale}", "cldr.dic"),
+			# Translators: The name of a symbols dictionary with data from the unicode CLDR.
+			displayName=_("Unicode Consortium data (including emoji)"),
+		),
+	)
+	_symbolDictionaryDefinitions.append(
+		SymbolDictionaryDefinition(
+			name="builtin",
+			path=os.path.join(globalVars.appDir, "locale", "{locale}", "symbols.dic"),
+			allowComplexSymbols=True,
+			mandatory=True,
+		),
+	)
+
+	import addonHandler
+
+	for addon in addonHandler.getRunningAddons():
+		try:
+			symbolsDict = addon.manifest.get("symbolDictionaries")
+			if not symbolsDict:
+				continue
+			log.debug(
+				f"Found {len(symbolsDict)} symbol dictionary entries in manifest for add-on {addon.name!r}",
+			)
+			directory = os.path.join(addon.path, "locale", "{locale}")
+			for name, dictConfig in symbolsDict.items():
+				definition = SymbolDictionaryDefinition(
+					name=name,
+					path=os.path.join(directory, f"symbols-{name}.dic"),
+					displayName=dictConfig["displayName"],
+					allowComplexSymbols=False,
+					mandatory=dictConfig["mandatory"],
+				)
+				if not definition.availableLocales:
+					log.error(f"No {name!r} symbol dictionary files found for add-on {addon.name!r}")
+					continue
+				_symbolDictionaryDefinitions.append(definition)
```

Comment:
can the scope of the try block could be reduced? I think lines 871-877 (from the try block, to the for loop), can be moved out safely.
Would it be safe to only call the try/except for each iteration of the for loop? that way other add-ons can still be loaded if the first fails

Reply:
I think I understand what you meant, hope this is better.